src/codecs/webp/extended.rs

```diff
@@ -0,0 +1,683 @@
+    duration: u32,
+    use_alpha_blending: bool,
+    dispose: bool,
+    image: WebPStatic,
+}
+
+pub(crate) fn read_extended_header<R: Read>(reader: &mut R) -> ImageResult<WebPExtendedInfo> {
+    let chunk_flags = reader.read_u8()?;
+
+    let reserved_first = chunk_flags & 0b11000000;
+    let icc_profile = chunk_flags & 0b00100000 != 0;
+    let alpha = chunk_flags & 0b00010000 != 0;
+    let exif_metadata = chunk_flags & 0b00001000 != 0;
+    let xmp_metadata = chunk_flags & 0b00000100 != 0;
+    let animation = chunk_flags & 0b00000010 != 0;
+    let reserved_second = chunk_flags & 0b00000001;
+
+    let mut reserved_third: [u8; 3] = [0; 3];
+    reader.read_exact(&mut reserved_third)?;
+
+    if reserved_first != 0 || reserved_second != 0 || reserved_third != [0, 0, 0] {
+        return Err(DecoderError::InfoBitsInvalid.into());
+    }
+
+    let canvas_width = read_3_bytes(reader)? + 1;
+    let canvas_height = read_3_bytes(reader)? + 1;
+
+    //product of canvas dimensions cannot be larger than u32 max
+    if u32::checked_mul(canvas_width, canvas_height).is_none() {
+        return Err(DecoderError::SizeMismatch.into());
+    }
+
+    let info = WebPExtendedInfo {
+        _icc_profile: icc_profile,
+        alpha,
+        _exif_metadata: exif_metadata,
+        _xmp_metadata: xmp_metadata,
+        _animation: animation,
+        canvas_width,
+        canvas_height,
+    };
+
+    Ok(info)
+}
+
+fn read_anim_frame<R: Read>(mut reader: R) -> ImageResult<AnimatedFrame> {
+    let frame_x = read_3_bytes(&mut reader)?;
+    let frame_y = read_3_bytes(&mut reader)?;
+
+    let frame_width = read_3_bytes(&mut reader)? + 1;
+    let frame_height = read_3_bytes(&mut reader)? + 1;
+    let duration = read_3_bytes(&mut reader)?;
+
+    let frame_info = reader.read_u8()?;
+    let reserved = frame_info & 0b11111100;
+    if reserved != 0 {
+        return Err(DecoderError::InfoBitsInvalid.into());
+    }
+    let use_alpha_blending = frame_info & 0b00000010 == 0;
+    let dispose = frame_info & 0b00000001 != 0;
+
+    //read normal bitstream now
+    let static_image = read_image(&mut reader, frame_width, frame_height)?;
+
+    let frame = AnimatedFrame {
+        offset_x: frame_x,
+        offset_y: frame_y,
+        width: frame_width,
+        height: frame_height,
+        duration,
+        use_alpha_blending,
+        dispose,
+        image: static_image,
+    };
+
+    Ok(frame)
+}
+
+fn read_3_bytes<R: Read>(reader: &mut R) -> ImageResult<u32> {
+    let mut buffer: [u8; 3] = [0; 3];
+    reader.read_exact(&mut buffer)?;
+    let value: u32 =
+        (u32::from(buffer[2]) << 16) | (u32::from(buffer[1]) << 8) | u32::from(buffer[0]);
+    Ok(value)
+}
+
+fn read_lossy<R: Read>(reader: &mut R) -> ImageResult<VP8Frame> {
+    let (cursor, chunk) =
+        read_chunk(reader)?.ok_or_else(|| Error::from(io::ErrorKind::UnexpectedEof))?;
+
+    if chunk != WebPRiffChunk::VP8 {
+        return Err(DecoderError::ChunkInvalid.into());
```

Comment:
Maybe this should be a separate error variant?

Reply:
Yeah I wasn't really sure about classifying errors, maybe could use the ChunkHeaderInvalid from decoder.rs?

Reply:
I wouldn't be too worried about code reuse here. Rather be as detailed and specific as useful for either certifying _how_  the file standard contradicts the specification (i.e. something you can check on your own) or for locating bugs in the decoder.

Reply:
More code duplication reuse can be helpful. It allows mapping specific error information to a line of code, if this information isn't directly visible from other information. In private projects I almost always include the line of code and even source files. We don't want that kind of information leakage though, but being cautious on reusing error variants already goes most of the way.

Reply:
I tried changing all the errors to be more specific to the issue with the image file. With that one, I did use the ChunkHeaderInvalid since I think it characterises the error received.